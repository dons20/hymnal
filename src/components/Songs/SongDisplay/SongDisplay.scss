.songs > .container {
	display: grid;
	grid-row: 1 / span 2;
	grid-auto-rows: max-content;
	justify-content: center;
	grid-template-columns: minmax(auto, var(--chakra-sizes-container-md));

	& > .body {
		margin: 20px auto;
		display: grid;
		white-space: pre-line;
		grid-row-gap: 2.5rem;
		// width: 100%;
		// max-width: var(--chakra-sizes-container-sm);
	}

	& > .header {
		align-items: center;
		gap: 15px;
		display: grid;
		font-weight: bold;
		margin: auto;
		max-width: 1000px;
		width: 100%;
	}

	& > .footer {
		justify-self: flex-end;
		font-style: italic;
	}

	& .label {
		display: block;
		font-style: italic;
		font-size: calc(10px + 1vmin);
		font-weight: 600;
	}

	@media only screen and (min-width: 550px) {
		& > .header {
			grid-template-columns: auto 1fr auto 50px;
		}
	}
}
