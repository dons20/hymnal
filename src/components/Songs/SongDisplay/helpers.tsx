import React from 'react';
import { Box, Text } from '@mantine/core';

/**
 * Helper function to render chorus with vocal parts and indicators
 */
export const renderChorusWithParts = (song: Song, isDark: boolean) => {
  if (!song.chorusParts || song.chorusParts.length === 0) {
    // Fallback to regular chorus rendering
    return <Text className="song-text">{song.chorus}</Text>;
  }

  return (
    <Box className="chorus-parts">
      {song.chorusParts.map((part, index) => (
        <Box
          key={index}
          className="chorus-part"
          style={{ position: 'relative', marginBottom: '0.5rem' }}
        >
          {part.part && (
            <Text
              component="span"
              className="part-indicator-regular"
              style={
                {
                  '--part-indicator-color':
                    part.part === 'soprano'
                      ? isDark
                        ? '#74c0fc'
                        : '#1c7ed6' // Blue for soprano
                      : isDark
                        ? '#ffec99'
                        : '#fd7e14', // Orange for baritone
                  '--part-indicator-bg':
                    part.part === 'soprano'
                      ? isDark
                        ? 'rgba(116, 192, 252, 0.15)'
                        : 'rgba(28, 126, 214, 0.08)'
                      : isDark
                        ? 'rgba(255, 236, 153, 0.15)'
                        : 'rgba(253, 126, 20, 0.08)',
                  '--part-indicator-border':
                    part.part === 'soprano'
                      ? isDark
                        ? 'rgba(116, 192, 252, 0.3)'
                        : 'rgba(28, 126, 214, 0.25)'
                      : isDark
                        ? 'rgba(255, 236, 153, 0.3)'
                        : 'rgba(253, 126, 20, 0.25)',
                } as React.CSSProperties
              }
            >
              {part.part === 'soprano' ? 'A' : part.part === 'baritone' ? 'B' : '♪'}
            </Text>
          )}
          <Text
            className="song-text"
            style={
              {
                paddingLeft: part.part ? '0' : '0',
                fontWeight: part.part === 'soprano' ? 500 : 400,
                '--part-text-color':
                  part.part === 'soprano'
                    ? isDark
                      ? '#e9ecef'
                      : '#212529' // Standard text color for soprano
                    : isDark
                      ? '#ced4da'
                      : '#495057', // Slightly muted for baritone
                color: 'var(--part-text-color)',
              } as React.CSSProperties
            }
          >
            {part.text}
          </Text>
        </Box>
      ))}
    </Box>
  );
};

/**
 * Helper function to render presentation slide content (verses or chorus with vocal parts)
 */
export const renderPresentationContent = (content: string | VocalPart[], isDark: boolean) => {
  if (typeof content === 'string') {
    return <Text className="presentation-text">{content}</Text>;
  }

  // Handle VocalPart[] for presentation mode with larger styling
  return (
    <Box className="presentation-content-container">
      {content.map((part, index) => {
        // Check if this is the end of a B part followed by an A part (new pair starting)
        const isEndOfPair =
          part.part === 'baritone' &&
          index < content.length - 1 &&
          content[index + 1]?.part === 'soprano';

        return (
          <React.Fragment key={index}>
            <Box
              className="presentation-part-container"
              style={{
                marginBottom: isEndOfPair ? '2rem' : '1.5rem',
              }}
            >
              {part.part && (
                <Text
                  component="span"
                  className="part-indicator-presentation"
                  style={
                    {
                      '--part-indicator-color':
                        part.part === 'soprano'
                          ? isDark
                            ? '#74c0fc'
                            : '#1c7ed6' // Blue for soprano
                          : isDark
                            ? '#ffec99'
                            : '#fd7e14', // Orange for baritone
                      '--part-indicator-bg':
                        part.part === 'soprano'
                          ? isDark
                            ? 'rgba(116, 192, 252, 0.15)'
                            : 'rgba(28, 126, 214, 0.08)'
                          : isDark
                            ? 'rgba(255, 236, 153, 0.15)'
                            : 'rgba(253, 126, 20, 0.08)',
                      '--part-indicator-border':
                        part.part === 'soprano'
                          ? isDark
                            ? 'rgba(116, 192, 252, 0.3)'
                            : 'rgba(28, 126, 214, 0.25)'
                          : isDark
                            ? 'rgba(255, 236, 153, 0.3)'
                            : 'rgba(253, 126, 20, 0.25)',
                    } as React.CSSProperties
                  }
                >
                  {part.part === 'soprano' ? 'A' : part.part === 'baritone' ? 'B' : '♪'}
                </Text>
              )}
              <Text
                className="presentation-part-text"
                style={
                  {
                    fontWeight: part.part === 'soprano' ? 600 : 400,
                    '--part-text-color':
                      part.part === 'soprano'
                        ? isDark
                          ? '#e9ecef'
                          : '#212529' // Standard text color for soprano
                        : isDark
                          ? '#ced4da'
                          : '#495057', // Slightly muted for baritone
                  } as React.CSSProperties
                }
              >
                {part.text}
              </Text>
            </Box>
            {isEndOfPair && (
              <Box
                className="pair-divider"
                style={
                  {
                    '--divider-color': isDark
                      ? 'rgba(255, 255, 255, 0.2)'
                      : 'rgba(0, 0, 0, 0.15)',
                  } as React.CSSProperties
                }
              />
            )}
          </React.Fragment>
        );
      })}
    </Box>
  );
};
