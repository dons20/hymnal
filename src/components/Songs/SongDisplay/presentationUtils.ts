import React from 'react';

export interface PresentationSlide {
  type: string;
  content: string | VocalPart[];
  label: string;
}

/**
 * Creates presentation slides in the correct order (V1 -> C -> V2 -> C -> V3 -> C)
 */
export const createPresentationSlides = (song: Song | null): PresentationSlide[] => {
  if (!song) return [];

  const slides: PresentationSlide[] = [];
  const hasChorus = Boolean(song.chorus);

  song.verse.forEach((verse: string, index: number) => {
    slides.push({
      type: 'verse',
      content: verse,
      label: `Verse ${index + 1}`,
    });

    // Add chorus after each verse
    if (hasChorus) {
      slides.push({
        type: 'chorus',
        content:
          song.chorusParts && song.chorusParts.length > 0 ? song.chorusParts : song.chorus,
        label: 'Chorus',
      });
    }
  });

  return slides;
};

/**
 * Touch event handler for swipe start
 */
export const createTouchStartHandler = (
  setTouchEnd: React.Dispatch<React.SetStateAction<{ x: number; y: number } | null>>,
  setTouchStart: React.Dispatch<React.SetStateAction<{ x: number; y: number } | null>>
) => {
  return (e: React.TouchEvent) => {
    setTouchEnd(null);
    setTouchStart({
      x: e.targetTouches[0].clientX,
      y: e.targetTouches[0].clientY,
    });
  };
};

/**
 * Touch event handler for swipe move
 */
export const createTouchMoveHandler = (
  setTouchEnd: React.Dispatch<React.SetStateAction<{ x: number; y: number } | null>>
) => {
  return (e: React.TouchEvent) => {
    setTouchEnd({
      x: e.targetTouches[0].clientX,
      y: e.targetTouches[0].clientY,
    });
  };
};

/**
 * Touch event handler for swipe end with navigation callbacks
 */
export const createTouchEndHandler = (
  touchStart: { x: number; y: number } | null,
  touchEnd: { x: number; y: number } | null,
  nextSlide: () => void,
  previousSlide: () => void,
  minSwipeDistance: number = 50
) => {
  return () => {
    if (!touchStart || !touchEnd) return;

    const distanceX = touchStart.x - touchEnd.x;
    const distanceY = touchStart.y - touchEnd.y;
    const isHorizontalSwipe = Math.abs(distanceX) > Math.abs(distanceY);

    if (isHorizontalSwipe && Math.abs(distanceX) > minSwipeDistance) {
      if (distanceX > 0) {
        nextSlide();
      } else {
        previousSlide();
      }
    }
  };
};

/**
 * Click handler for directional navigation in presentation mode
 */
export const createPresentationClickHandler = (
  nextSlide: () => void,
  previousSlide: () => void
) => {
  return (e: React.MouseEvent<HTMLElement>) => {
    const target = e.currentTarget;
    const rect = target.getBoundingClientRect();
    const x = e.clientX - rect.left;
    const width = rect.width;

    const leftAreaEnd = width * 0.25;

    if (x < leftAreaEnd) {
      previousSlide();
    } else {
      nextSlide();
    }
  };
};
