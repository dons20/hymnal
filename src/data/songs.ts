export const version = "1.1.0";
export const SongsDB = [
	{
		number: 1,
		title: "There Was One Who Was Willing",
		verse: [
			"There was one who was willing to die in my stead,\nThat a soul so unworthy might live,\nAnd the path to the cross He was willing to tread,\nAll the sins of my life to forgive.",
			"He is tender and loving and patient with me,\nWhile He cleanses my heart of its dross,\nBut there's no condemnation, I know I am free,\nFor my sins are all nailed to the cross.",
			"I will cling to my Saviour and never depart,\nI will joyfully journey each day,\nWith a song on my lips and a song in my heart,\nThat my sins have been taken away.",
		],
		chorus: "They are nailed to the Cross, they are nailed to the cross,\nO how much He was willing to bear!\nWith what anguish and loss, Jesus went to the cross!\nBut He carried my sins with Him there.",
		author: "James Montgomery",
	},
	{
		number: 2,
		title: "Redeemed! How I Love To Proclaim It!",
		verse: [
			"Redeemed! How I love to proclaim it!\nRedeemed by the blood of the Lamb;\nRedeemed through His infinite mercy,\nHis child, and forever, I am.",
			"Redeemed! And so happy in Jesus!\nNo language my rapture can tell;\nI know that the light of His presence\nWith me doth continually dwell.",
			"I think of my blessed Redeemer,\nI think of Him all the day long;\nI sing, for I cannot be silent;\nHis love is the theme of my song.",
			"I know I shall see in his beauty\nThe King in whose law I delight,\nWho lovingly guardeth my footsteps,\nAnd giveth me songs in the night.",
			"I know there's a crown that is waiting\nIn yonder bright mansion for me;\nAnd soon, with the spirits made perfect,\nAt home with the Lord I shall be.",
		],
		chorus: "Redeemed, redeemed, redeemed by the blood of the Lamb;\nRedeemed, redeemed, His child and forever, I am.",
		author: "Fanny J. Crosby",
	},
	{
		number: 3,
		title: "O Now I See The Crimson Wave",
		verse: [
			"O now I see the crimson wave, the fountain deep and wide;\nJesus, my Lord, mighty to save, points to His wounded side.",
			"I see the new creation rise, I hear the speaking blood;\nIt speaksâ€”polluted nature dies, sinks 'neath the cleansing flood.",
			"I rise to walk in heaven's own light, above the world and sin;\nWith heart made pure and garments white, and Christ enthroned within.",
			"Amazing grace! 'tis heaven below to feel the blood applied,\nAnd Jesus, only Jesus, know, my Jesus crucified.",
		],
		chorus: "The cleansing stream I see, I see,\nI plunge, and O, it cleanseth me!\nO praise the Lord! It cleanseth me,\nIt cleanseth me, yes, cleanseth me.",
	},
	{
		number: 4,
		title: "Dying With Jesus",
		verse: [
			"Dying with Jesus, by death reckoned mine,\nLiving with Jesus a new life divine,\nLooking to Jesus till glory doth shine,\nMoment by moment O Lord, I am Thine.",
			"Never a trial that He is not there,\nNever a burden that He doth not bear,\nNever a sorrow that He doth not share,\nMoment by moment I'm under His care.",
			"Never a heartache, and never a groan,\nNever a teardrop and never a moan;\nNever a danger but there on the throne,\nMoment by moment He thinks of His own.",
			"Never a weakness that He doth not feel,\nNever a sickness that He cannot heal;\nMoment by moment, in woe or in weal,\nJesus, my Saviour, abides with me still.",
		],
		chorus: "Moment by moment I'm kept in His love;\nMoment by moment I've life from above;\nLooking to Jesus till glory doth shine;\nMoment by moment, O Lord, I am Thine.",
		author: "D. W. Whittle",
	},
	{
		number: 5,
		title: "Come, Every Soul By Sin Oppressed",
		verse: [
			"Come, every soul by sin oppressed, there's mercy with the Lord,\nAnd He will surely give you rest, by trusting in His word.",
			"For Jesus shed His precious blood rich blessings to bestow;\nPlunge now into the crimson flood that washes white as snow.",
			"Yes, Jesus is the truth, the way, that leads you into rest;\nBelieve in Him without delay, and you are fully blest.",
			"Come, then, and join this holy band, and on to glory go,\nTo dwell in that celestial land, where joys immortal flow.",
		],
		chorus: "Only trust Him, only trust Him, only trust Him now;\nHe will save you, He will save you, He will save you now.",
		author: "J. H. Stockton",
	},
];
