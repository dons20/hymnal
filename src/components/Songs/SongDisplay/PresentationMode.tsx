import React from 'react';
import { Helmet } from '@dr.pogodin/react-helmet';
import { ActionIcon, Box, Container, Group, Portal, Text, Transition } from '@mantine/core';
import { renderPresentationContent } from './helpers';
import { PresentationSlide } from './presentationUtils';

interface PresentationModeProps {
  presentationMode: boolean;
  presentationSlides: PresentationSlide[];
  currentSlide: number;
  slideTransitioning: boolean;
  isDark: boolean;
  reducedMotion: boolean;
  songTitle: string;
  onClose: () => void;
  onClick: (e: React.MouseEvent<HTMLElement>) => void;
  onTouchStart: (e: React.TouchEvent) => void;
  onTouchMove: (e: React.TouchEvent) => void;
  onTouchEnd: () => void;
  onSlideClick: (index: number) => void;
}

export const PresentationMode: React.FC<PresentationModeProps> = ({
  presentationMode,
  presentationSlides,
  currentSlide,
  slideTransitioning,
  isDark,
  reducedMotion,
  songTitle,
  onClose,
  onClick,
  onTouchStart,
  onTouchMove,
  onTouchEnd,
  onSlideClick,
}) => {
  if (!presentationMode || presentationSlides.length === 0) return null;

  const currentSlideData = presentationSlides[currentSlide];

  return (
    <>
      {/* Render regular content but hidden */}
      <div style={{ display: 'none' }}>
        <Container className="container" size="lg" my="md" py="lg" px="xl">
          <Helmet>
            <title>{`Hymns for All Times | ${songTitle}`}</title>
          </Helmet>
        </Container>
      </div>

      {/* Portal for presentation mode - rendered outside component tree */}
      <Portal>
        <Transition
          mounted={presentationMode}
          transition="fade"
          duration={reducedMotion ? 0 : 300}
        >
          {(styles) => (
            <Box
              className="presentation-mode-overlay presentation-mode"
              bg={isDark ? 'gray.9' : 'white'}
              style={{
                ...styles,
              }}
              onClick={onClick}
              onTouchStart={onTouchStart}
              onTouchMove={onTouchMove}
              onTouchEnd={onTouchEnd}
            >
              {/* Close button - positioned safely at top right */}
              <ActionIcon
                className="presentation-close-button"
                variant="outline"
                size="lg"
                color="red"
                onClick={(e) => {
                  e.stopPropagation();
                  onClose();
                }}
              >
                âœ•
              </ActionIcon>

              {/* Slide content with transition */}
              <Transition
                mounted={!slideTransitioning}
                transition="fade"
                duration={reducedMotion ? 0 : 150}
              >
                {(contentStyles) => (
                  <Box
                    className="presentation-slide-content"
                    style={{
                      ...contentStyles,
                    }}
                  >
                    <Text
                      className="presentation-slide-label"
                      size="lg"
                      fw={600}
                      mb="lg"
                      c={isDark ? 'blue.4' : 'blue.6'}
                    >
                      {currentSlideData.label}
                    </Text>
                    {renderPresentationContent(currentSlideData.content, isDark)}
                  </Box>
                )}
              </Transition>

              {/* Navigation indicators - positioned safely at bottom */}
              <Group
                className="presentation-indicators"
                style={
                  {
                    '--indicators-bg': isDark ? 'rgba(0, 0, 0, 0.3)' : 'rgba(255, 255, 255, 0.8)',
                    '--indicators-border': isDark
                      ? 'rgba(255, 255, 255, 0.1)'
                      : 'rgba(0, 0, 0, 0.1)',
                  } as React.CSSProperties
                }
              >
                {presentationSlides.map((_, index) => (
                  <Box
                    key={index}
                    className="presentation-indicator-dot"
                    style={
                      {
                        '--dot-color':
                          index === currentSlide
                            ? 'var(--mantine-color-blue-5)'
                            : isDark
                              ? 'var(--mantine-color-gray-5)'
                              : 'var(--mantine-color-gray-4)',
                      } as React.CSSProperties
                    }
                    onClick={(e) => {
                      e.stopPropagation();
                      onSlideClick(index);
                    }}
                  />
                ))}
              </Group>
            </Box>
          )}
        </Transition>
      </Portal>
    </>
  );
};
