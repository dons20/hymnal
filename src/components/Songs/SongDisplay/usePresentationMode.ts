import { useEffect, useMemo, useState } from 'react';
import { createPresentationSlides, PresentationSlide } from './presentationUtils';

interface UsePresentationModeProps {
  song: Song | null;
  reducedMotion: boolean;
}

interface UsePresentationModeReturn {
  presentationMode: boolean;
  currentSlide: number;
  slideTransitioning: boolean;
  presentationSlides: PresentationSlide[];
  touchStart: { x: number; y: number } | null;
  touchEnd: { x: number; y: number } | null;
  setTouchStart: React.Dispatch<React.SetStateAction<{ x: number; y: number } | null>>;
  setTouchEnd: React.Dispatch<React.SetStateAction<{ x: number; y: number } | null>>;
  togglePresentationMode: () => void;
  closePresentationMode: () => void;
  nextSlide: () => void;
  previousSlide: () => void;
  goToSlide: (index: number) => void;
}

/**
 * Custom hook to manage presentation mode state and navigation
 */
export const usePresentationMode = ({
  song,
  reducedMotion,
}: UsePresentationModeProps): UsePresentationModeReturn => {
  const [presentationMode, setPresentationMode] = useState(false);
  const [currentSlide, setCurrentSlide] = useState(0);
  const [slideTransitioning, setSlideTransitioning] = useState(false);
  const [touchStart, setTouchStart] = useState<{ x: number; y: number } | null>(null);
  const [touchEnd, setTouchEnd] = useState<{ x: number; y: number } | null>(null);

  // Create presentation slides
  const presentationSlides = useMemo(() => createPresentationSlides(song), [song]);

  const togglePresentationMode = () => {
    setPresentationMode(!presentationMode);
    setCurrentSlide(0);
  };

  const closePresentationMode = () => {
    setPresentationMode(false);
    setCurrentSlide(0);
    setSlideTransitioning(false);
  };

  const nextSlide = () => {
    if (reducedMotion) {
      if (currentSlide < presentationSlides.length - 1) {
        setCurrentSlide(currentSlide + 1);
      } else {
        setCurrentSlide(0);
      }
    } else {
      setSlideTransitioning(true);
      setTimeout(() => {
        if (currentSlide < presentationSlides.length - 1) {
          setCurrentSlide(currentSlide + 1);
        } else {
          setCurrentSlide(0);
        }
        setSlideTransitioning(false);
      }, 150);
    }
  };

  const previousSlide = () => {
    if (reducedMotion) {
      if (currentSlide > 0) {
        setCurrentSlide(currentSlide - 1);
      } else {
        setCurrentSlide(presentationSlides.length - 1);
      }
    } else {
      setSlideTransitioning(true);
      setTimeout(() => {
        if (currentSlide > 0) {
          setCurrentSlide(currentSlide - 1);
        } else {
          setCurrentSlide(presentationSlides.length - 1);
        }
        setSlideTransitioning(false);
      }, 150);
    }
  };

  const goToSlide = (index: number) => {
    if (!reducedMotion) {
      setSlideTransitioning(true);
      setTimeout(() => {
        setCurrentSlide(index);
        setSlideTransitioning(false);
      }, 150);
    } else {
      setCurrentSlide(index);
    }
  };

  // ESC key handler for presentation mode
  useEffect(() => {
    const handleEscKey = (event: KeyboardEvent) => {
      if (event.key === 'Escape' && presentationMode) {
        closePresentationMode();
      }
    };

    if (presentationMode) {
      document.addEventListener('keydown', handleEscKey);
    }

    return () => {
      document.removeEventListener('keydown', handleEscKey);
    };
  }, [presentationMode]);

  // Arrow key navigation for presentation mode
  useEffect(() => {
    const handleArrowKeys = (event: KeyboardEvent) => {
      if (!presentationMode) return;

      if (event.key === 'ArrowRight') {
        event.preventDefault();
        nextSlide();
      } else if (event.key === 'ArrowLeft') {
        event.preventDefault();
        previousSlide();
      }
    };

    if (presentationMode) {
      document.addEventListener('keydown', handleArrowKeys);
    }

    return () => {
      document.removeEventListener('keydown', handleArrowKeys);
    };
  }, [presentationMode, currentSlide, presentationSlides.length, reducedMotion]);

  return {
    presentationMode,
    currentSlide,
    slideTransitioning,
    presentationSlides,
    touchStart,
    touchEnd,
    setTouchStart,
    setTouchEnd,
    togglePresentationMode,
    closePresentationMode,
    nextSlide,
    previousSlide,
    goToSlide,
  };
};
