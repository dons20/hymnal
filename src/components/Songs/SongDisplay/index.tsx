import React, { useEffect, useMemo } from "react";
import { FaHeart, FaArrowCircleRight, FaArrowCircleLeft, FaBook } from "react-icons/fa";
import { Box, Container, Text, IconButton, SimpleGrid, Icon } from "@chakra-ui/react";
import { Navigate, useNavigate, useParams } from "react-router-dom";
import { useColorModeValue } from "@chakra-ui/color-mode";
import { useMainContext } from "utils/context";
import { updateFavesDB } from "helpers";
import { Helmet } from "react-helmet";
import { Button } from "components";
import "./SongDisplay.scss";

type ParamTypes = {
	songID?: string;
};

function SongDisplay() {
	const navigate = useNavigate();
	const { songID } = useParams<ParamTypes>();
	const { songs, favourites, setFavourites, dispatch } = useMainContext();
	const authorColor = useColorModeValue("#555555", "gray.300");
	const songBG = useColorModeValue("gray.50", "inherit");
	const songShadow = useColorModeValue("md", undefined);
	const modalBG = useColorModeValue("gray.100", "gray.800");
	const favActiveIconColor = useColorModeValue("var(--chakra-colors-red-500)", "var(--chakra-colors-red-300)");
	const favIconColor = useColorModeValue("var(--chakra-colors-gray-600)", "var(--chakra-colors-gray-500)");
	const favActiveIconBG = useColorModeValue("var(--chakra-colors-red-50)", "");
	const songIndex = parseInt(songID || "1", 10) - 1;
	const songToRender = songs.find(song => song.number === songIndex + 1) || null;

	// Ensure the route parameter is a number before rendering anything
	useEffect(() => {
		if (!/\d+/.test(songID!)) navigate(-1);
	}, [songID, navigate]);

	useEffect(() => {
		if (songs.length > 1) dispatch!({ type: "setTitle", payload: songToRender?.title || "" });
	}, [dispatch, songs, songToRender]);

	const songBody = useMemo(
		() =>
			songs.length > 1 &&
			React.Children.toArray(
				songToRender?.verse.map((verse, i) => {
					if (i === 1 && songToRender.chorus) {
						return (
							<>
								<Box className="chorus">
									<span className="label">Chorus</span>
									{songToRender.chorus}
								</Box>
								<Box className="verse">
									<span className="label">Verse {i + 1}</span>
									{verse}
								</Box>
							</>
						);
					}

					return (
						<Box className="verse">
							<span className="label">Verse {i + 1}</span>
							{verse}
						</Box>
					);
				})
			),
		[songs, songToRender]
	);

	if (songToRender === null) return <Navigate to="../index" replace />;

	const isFavourite = favourites.includes(songToRender.number - 1);
	const isFirstSong = songToRender.number > 1;
	const isLastSong = songToRender.number < songs.length - 1;

	const toggleFavourite = (number: number) => {
		let faves = [];
		if (favourites.includes(number - 1)) {
			faves = favourites.filter(fave => fave !== number - 1);
			setFavourites(faves);
		} else {
			faves = [...favourites, number - 1];
			setFavourites(faves);
		}
		updateFavesDB(faves);
	};

	// Useful for navigating to the correct spot in the song list
	const backToIndex = () => navigate("../index", { state: { songNumber: songToRender.number - 1 } });
	const previousSong = () => navigate(`../${songToRender.number - 1}`);
	const nextSong = () => navigate(`../${songToRender.number + 1}`);

	return (
		<Container
			className="container"
			maxW="container.lg"
			bg={songBG}
			shadow={songShadow}
			my={4}
			py="1rem"
			px="1.5rem"
		>
			{/* @ts-expect-error Helmet no longer updated */}
			<Helmet>
				<title>{`Hymns for All Times | ${songToRender!.title}`}</title>
			</Helmet>
			<Box className="header" pos="relative" pr="5">
				<Text># {songToRender!.number}</Text>
				<Text>{songToRender!.title}</Text>
				<Box>
					<IconButton
						colorScheme="gray"
						icon={<FaBook color={favIconColor} />}
						onClick={backToIndex}
						_hover={{ shadow: "md" }}
						bgColor={modalBG}
						aria-label="Back to songs index"
						variant="outline"
						size="lg"
						maxW="60px"
						pr={1}
						mr={5}
					/>
					<IconButton
						colorScheme={isFavourite ? "red" : "gray"}
						bgColor={isFavourite ? favActiveIconBG : modalBG}
						_hover={{ shadow: "md" }}
						icon={<FaHeart color={isFavourite ? favActiveIconColor : favIconColor} />}
						aria-label="Add to Favourites"
						size="lg"
						variant="outline"
						className="faveIcon"
						onClick={() => toggleFavourite(songToRender.number)}
						maxW="60px"
					/>
				</Box>
			</Box>
			<Box className="body">{songBody}</Box>
			{songToRender.author && (
				<Text className="footer" color={authorColor}>
					{songToRender.author}
				</Text>
			)}
			<SimpleGrid columns={{ sm: isFirstSong || isLastSong ? 1 : 2 }} justifyContent={{ md: "space-around" }} mb={5} mt={5} spacing={5}>
				{isFirstSong && (
					<Button onClick={previousSong} flex="1">
						<Icon as={FaArrowCircleLeft} size={20} mr={3} /> Previous Song
					</Button>
				)}
				{isLastSong && (
					<Button onClick={nextSong} flex="1">
						Next Song <Icon as={FaArrowCircleRight} size={20} ml={3} />
					</Button>
				)}
			</SimpleGrid>
			<Button onClick={backToIndex} variant="outline">
				Back to Index
			</Button>
		</Container>
	);
}

export default SongDisplay;
