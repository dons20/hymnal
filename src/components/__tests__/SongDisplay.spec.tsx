import { Suspense } from "react";
import { render, screen } from "@testing-library/react";
import SongDisplay from "components/Songs/SongDisplay";
import { MemoryRouter } from "react-router-dom";
import { Providers } from "helpers/tests";
import { SongsDB } from "data/songs";

describe("#SongDisplay", () => {
	it("renders the matching song based on the route", () => {
		const songs = SongsDB;
		const favourites: Song[] = [];
		const dispatch = jest.fn();
		const { asFragment } = render(
			<MemoryRouter initialEntries={["/songs/1"]}>
				<Providers value={{ songs, favourites, dispatch }}>
					<Suspense fallback="">
						<SongDisplay />
					</Suspense>
				</Providers>
			</MemoryRouter>
		);
		expect(asFragment()).toMatchSnapshot();
		expect(screen.getByText(songs[0].title)).toBeInTheDocument();
	});
});
