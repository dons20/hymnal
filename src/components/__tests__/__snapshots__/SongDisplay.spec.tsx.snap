// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#SongDisplay renders the matching song based on the route 1`] = `
<DocumentFragment>
  <div
    class="chakra-container container css-t3somj"
  >
    <div
      class="header css-53le5g"
    >
      <p
        class="chakra-text css-0"
      >
        # 1
      </p>
      <p
        class="chakra-text css-0"
      >
        There Was One Who Was Willing
      </p>
      <div
        class="css-0"
      >
        <button
          aria-label="Back to songs index"
          class="chakra-button css-1uztc5k"
          type="button"
        >
          <svg
            aria-hidden="true"
            color="var(--chakra-colors-gray-600)"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M448 360V24c0-13.3-10.7-24-24-24H96C43 0 0 43 0 96v320c0 53 43 96 96 96h328c13.3 0 24-10.7 24-24v-16c0-7.5-3.5-14.3-8.9-18.7-4.2-15.4-4.2-59.3 0-74.7 5.4-4.3 8.9-11.1 8.9-18.6zM128 134c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm0 64c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm253.4 250H96c-17.7 0-32-14.3-32-32 0-17.6 14.4-32 32-32h285.4c-1.9 17.1-1.9 46.9 0 64z"
            />
          </svg>
        </button>
        <button
          aria-label="Add to Favourites"
          class="chakra-button faveIcon css-11itnfl"
          type="button"
        >
          <svg
            aria-hidden="true"
            color="var(--chakra-colors-gray-600)"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="body css-0"
    >
      <div
        class="verse css-0"
      >
        <span
          class="label"
        >
          Verse 1
        </span>
        There was one who was willing to die in my stead,
That a soul so unworthy might live,
And the path to the cross He was willing to tread,
All the sins of my life to forgive.
      </div>
      <div
        class="chorus css-0"
      >
        <span
          class="label"
        >
          Chorus
        </span>
        They are nailed to the Cross, they are nailed to the cross,
O how much He was willing to bear!
With what anguish and loss, Jesus went to the cross!
But He carried my sins with Him there.
      </div>
      <div
        class="verse css-0"
      >
        <span
          class="label"
        >
          Verse 2
        </span>
        He is tender and loving and patient with me,
While He cleanses my heart of its dross,
But there's no condemnation, I know I am free,
For my sins are all nailed to the cross.
      </div>
      <div
        class="verse css-0"
      >
        <span
          class="label"
        >
          Verse 3
        </span>
        I will cling to my Saviour and never depart,
I will joyfully journey each day,
With a song on my lips and a song in my heart,
That my sins have been taken away.
      </div>
    </div>
    <p
      class="chakra-text footer css-1etkgh2"
    >
      James Montgomery
    </p>
    <div
      class="css-mm4t09"
    >
      <button
        class="chakra-button css-1skxxy0"
        type="button"
      >
        Next Song 
        <svg
          class="chakra-icon css-11gi69r"
          fill="currentColor"
          focusable="false"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
          />
        </svg>
      </button>
    </div>
    <button
      class="chakra-button css-4xx2wk"
      type="button"
    >
      Back to Index
    </button>
  </div>
</DocumentFragment>
`;
